### PR TITLE
Master web setting dashboard improve ui mpa

### PR DESCRIPTION
**Pull request by hmo-odoo**
Task: https://www.odoo.com/web#id=32804&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720

Pad: https://pad.odoo.com/p/r.a4ccd6ba9e65fd86eaf44b1e1e39a442

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_settings_dashboard/static/src/js/dashboard.js [hunk 1]
```diff
@@ -76,12 +76,37 @@ var DashboardInvitations = Widget.extend({
...
+        this.s2input = this.parent.$('textarea#user_emails').select2({
+            placeholder: _t("Enter e-mail addresses"),
+            tags: [],
+            tokenSeparators: [" ", ","],
+            dropdownCss: {display:'none'},
+            width: '100%',
+        });
+        this.s2input.on("select2-open", function(e){
+            $('.select2-drop-mask').remove();
+        });
```
Comment (dmo-odoo) [inline addons/web_settings_dashboard/static/src/js/dashboard.js]: Why is this handler needed ? I removed it and couldn't find a difference.

File: addons/web_settings_dashboard/static/src/js/dashboard.js [hunk 2]
```diff
@@ -76,12 +76,37 @@ var DashboardInvitations = Widget.extend({
...
+    },
+    on_focusout: function(e){
+        var self = this;
+        var value = $(e.currentTarget).val();
+        this.s2input = this.parent.$('textarea#user_emails');
+        if(value){
+            var data = _.union(self.s2input.select2('data'), [{id:value, text:value}]);
+            self.s2input.select2('data', data);
+        }
+    },
```
Comment (dmo-odoo) [inline addons/web_settings_dashboard/static/src/js/dashboard.js]: Same here, why is this handler needed ? I removed it and couldn't find a difference.

File: addons/web_settings_dashboard/static/src/less/dashboard.less [hunk 3]
```diff
@@ -60,8 +60,11 @@
                     width: 100%;
                 }
             }
-            a.o_web_settings_dashboard_user {
-              line-height: 24px;
+            .badge {
+                    border-color: #d9d9d9;
+                    background-color: #d9d9d9;
+                    color: #333333;
+                    font-weight: normal;
```
Comment (dmo-odoo) [inline addons/web_settings_dashboard/static/src/less/dashboard.less]: Replace #d9d9d9 by @gray-lighter-dark and #333333 by @gray-dark

File: addons/web_settings_dashboard/static/tests/dashboard_tests.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            'pending_counts': 1,
+            'pending_users':[{'id': 1, 'email': 'xyz@odoo.com'}],
+        }
+
+        var parent = createParent({
+            'data': {'apps':{},
+                    'share':{},
+                    'users_info': self.data,
+                    }
+        });
```
Comment (dmo-odoo) [inline addons/web_settings_dashboard/static/tests/dashboard_tests.js]: Why do you need all this tinkering with createParent ? Can't you just define the data object above then do var parent = new Dashboard.Dashboard(); then have the mockRPC directly?

File: addons/web_settings_dashboard/static/tests/dashboard_tests.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            },
+        });
+
+        dashboard_invitations.appendTo($('#qunit-fixture'));
+        var inputbox = dashboard_invitations.$el.find('#user_emails').val("abc@odoo.com");
+        var event = jQuery.Event("keypress");
+        event.keyCode = 13;
+        if(inputbox.trigger(event)) {
+            dashboard_invitations.$el.find('.o_web_settings_dashboard_invitations').trigger('click');
+            assert.strictEqual(dashboard_invitations.data['pending_users'].length, 2, "New user created");
```
Comment (dmo-odoo) [inline addons/web_settings_dashboard/static/tests/dashboard_tests.js]: Might be interesting to add a another mock test that would intercept the create on res.users ?

File: addons/web_settings_dashboard/static/src/js/dashboard.js (hunk 2 above)
Comment (mpa-odoo) [reply addons/web_settings_dashboard/static/src/js/dashboard.js]: Actually removing this handler will produce the aforesaid bug.
Basically idea behind this code is to get the textbox value in select2 when clicking on 'invite' button right after typing email address in textbox.

File: addons/web_settings_dashboard/static/src/js/dashboard.js (hunk 1 above)
Comment (mpa-odoo) [reply addons/web_settings_dashboard/static/src/js/dashboard.js]: Actually removing this handler will produce the aforesaid bug.

**Comment by dmo-odoo** [comment]
I think using select2 here is not a good idea since you need to do additional hacks to make it work the way we want.

I suggest we do the following (which is closer to the spec anyway):
- template: replace the textarea with an input
- js:
    - in init: add this.emailsToInvite = []
    - add a listener on the input, listen for ',' and enter
    - when one of the above is entered, call the new "storeEmail" function which does:
        - check if the email is valid (you will need to move the code that tests the validity of emails into its own function so we can call it from here)
        - if it is invalid: this.do_warn(_t("Please provide valid email addresses"), "");
        - if it is valid:
            - add the email currently in the input to the emailsToInvite list
            - empty the input and add a badge with the new email above the input
            - the badge should include a fa-close button to remove the email from emailsToInvite and remove its associated badge
        - in send_invitations:
            - start by calling storeEmail if there is an email address in the input
            - retrieve the emails in self.emailsToInvite rather than from the textarea like before.

**Comment by hmo-odoo** [comment]
continue work on PR: https://github.com/odoo/odoo/pull/19531

**Pull request closed without merging**